swap/swap.go

```diff
@@ -434,86 +436,86 @@ func (s *Swap) Balances() (map[enode.ID]int64, error) {
-			var peerCheque Cheque
-			err = json.Unmarshal(value, &peerCheque)
-			if err == nil {
-				cheques[peer] = &peerCheque
-			}
-		}
-		return stop, err
+	// get peer cheques from store
+	err := s.addStoreCheques(sentChequePrefix, sentChequeResponseKey, cheques)
+	if err != nil {
+		return nil, err
 	}
-	err := s.store.Iterate(sentChequePrefix, chequesIterFunction)
+	err = s.addStoreCheques(receivedChequePrefix, receivedChequeResponseKey, cheques)
 	if err != nil {
 		return nil, err
 	}
 
-	return cheques, nil
-}
-
-// ReceivedCheque returns the last received cheque for a given peer
-func (s *Swap) ReceivedCheque(peer enode.ID) (cheque *Cheque, err error) {
-	if swapPeer := s.getPeer(peer); swapPeer != nil {
-		return swapPeer.getLastReceivedCheque(), nil
+	// fill in result with missing cheques
+	for _, peerCheques := range cheques {
+		// add nil as type of cheque if not present
+		if _, peerHasReceivedCheque := peerCheques[receivedChequeResponseKey]; !peerHasReceivedCheque {
+			peerCheques[receivedChequeResponseKey] = nil
+		}
+		if _, peerHasSentCheque := peerCheques[sentChequeResponseKey]; !peerHasSentCheque {
+			peerCheques[sentChequeResponseKey] = nil
+		}
 	}
-	err = s.store.Get(receivedChequeKey(peer), &cheque)
-	return cheque, err
-}
-
-// ReceivedCheques returns the last received cheques for all known SWAP peers
-func (s *Swap) ReceivedCheques() (map[enode.ID]*Cheque, error) {
-	cheques := make(map[enode.ID]*Cheque)
 
-	s.peersLock.Lock()
-	for peer, swapPeer := range s.peers {
-		swapPeer.lock.Lock()
-		cheques[peer] = swapPeer.getLastReceivedCheque()
-		swapPeer.lock.Unlock()
-	}
-	s.peersLock.Unlock()
+	return cheques, nil
+}
 
-	// add store cheques, if peer was not already added
+// add disk cheques for peers not already present in given cheques map
+func (s *Swap) addStoreCheques(chequePrefix string, chequeKey string, cheques map[enode.ID]map[string]*Cheque) error {
 	chequesIterFunction := func(key []byte, value []byte) (stop bool, err error) {
-		peer := keyToID(string(key), receivedChequePrefix)
-		if _, peerHasCheque := cheques[peer]; !peerHasCheque {
+		peer := keyToID(string(key), chequePrefix)
+		// make map if peer has no cheques entry yet
+		if _, peerHasCheques := cheques[peer]; !peerHasCheques {
+			cheques[peer] = make(map[string]*Cheque)
+		}
+		// add cheque from store if not already in result
+		if peerCheque := cheques[peer][chequeKey]; peerCheque == nil {
 			var peerCheque Cheque
 			err = json.Unmarshal(value, &peerCheque)
 			if err == nil {
-				cheques[peer] = &peerCheque
+				cheques[peer][chequeKey] = &peerCheque
 			}
 		}
 		return stop, err
 	}
-	err := s.store.Iterate(receivedChequePrefix, chequesIterFunction)
-	if err != nil {
-		return nil, err
+	return s.store.Iterate(chequePrefix, chequesIterFunction)
+}
+
+// PeerCheques returns the last sent and received cheques for a given peer
+func (s *Swap) PeerCheques(peer enode.ID) (map[string]*Cheque, error) {
```

Comment:
if you're just returning two `*Cheques` then please change the function signature to:
`func ....() (lastSentCheque, lastReceivedCheque *Cheque, err error) {`
using a hardcoded map here seems overly redundant

Reply:
i would like to, but i don't think i can. the purpose of `PeerCheques` is to serve as a remote process call. therefore i can only return 1 variable and 1 error.

Reply:
well then if it can return only 1 variable and 1 error, maybe split the endpoint to `LastSentPeerCheque(peer enode.ID)` and `LastReceivedPeerCheque(peer enode.ID)`?

Reply:
From the function body, there is no sign that last sent and last received cheque are in any relation and the do not need to be read under a single lock, so I do not see why there would not be two methods instead of one. If the usage is that both cheques are needed in a single rpc call, I would rather use a struct with explicit fields then a map.

Reply:
thanks for the input guys.

originally, i had implemented multiple calls as you two suggest here. we had [this discussion](https://github.com/ethersphere/swarm/pull/1863#discussion_r333861141) with @Eknir and @ralph-pichler and arrived at the conclusion that the swap API would be better off with a couple of basic methods, rather than 5 or 6 very similar ones. 

> i am still voting for having two endpoints. i don't think that having two calls instead of one is a big overhead if it contributes towards a more maintainable and readable code. @janos, @nonsense your opinions too would be appreciated.

two endpoints will result in essentially duplicated code. the functions will be very very similar—their only difference being in a store key, basically. this carries over to the tests, and these are already bloated files.

so if it's a matter of maintainability, i am now thinking splitting this function is not the best option.

if it's a matter of readability, i am all for returning a `struct` as @janos suggests. in fact, i've added this to the code.

> From the function body, there is no sign that last sent and last received cheque are in any relation and the do not need to be read under a single lock, so I do not see why there would not be two methods instead of one. If the usage is that both cheques are needed in a single rpc call, I would rather use a struct with explicit fields then a map.

not sure what you mean with this lock you mention, but yes, the relationship between these cheques is more in the business logic side of it, rather than the code.

Reply:
> originally, i had implemented multiple calls as you two suggest here. we had this discussion with @Eknir and @ralph-pichler and arrived at the conclusion that the swap API would be better off with a couple of basic methods, rather than 5 or 6 very similar ones.

Well then, it's good you have more peer reviews, no? I personally wouldn't rubber stamp this PR in this condition. I find these deep equal assertions in the test cases looking like something coming from a Javascript test suite rather than a Golang one.

> two endpoints will result in essentially duplicated code

can you please say how much duplicate code we're talking about? is it 2 lines? 10? 100?

> so if it's a matter of maintainability, i am now thinking splitting this function is not the best option.

Can you give any reasons to _why_ you think this is not the best option?

> not sure what you mean with this lock you mention

He means that you're not holding any joint lock since the atomicity of their retrieval is supposed to be guaranteed by the state store rather than the `SWAP` code.

Reply:
> Well then, it's good you have more peer reviews, no? I personally wouldn't rubber stamp this PR in this condition. I find these deep equal assertions in the test cases looking like something coming from a Javascript test suite rather than a Golang one.

can you please point out what is wrong with the deep equal assertions? do you think we should have a custom method to compare these structs?

they were originally `map` variables, which is why `DeepEqual` was used when comparing them in the first place. they were then changed to `struct` variables, to improve readability. 

a `==` comparison won't work here, so we either use `reflect.DeepEqual` or a custom function to compare them.
 
> can you please say how much duplicate code we're talking about? is it 2 lines? 10? 100?

[here](https://codeshare.io/GLpnvb) is how the 2 endpoints implementation would look.

- `ReceivedCheque` is a clone of `SentCheque` (15 lines)
- `TestReceivedCheque` is a clone of `TestSentCheque` (86 lines)

for comparison, [here](https://codeshare.io/2W9Njv) is the current implementation.

i think it's more an issue of best practices rather than # of lines, but there they are.

> Can you give any reasons to _why_ you think this is not the best option?

in this case, i don't think copy-pasting a function and changing 1 key improves maintainability. 

this applies even more-so to tests, which are lengthier. any changes done to either the sent or received cheque function will _always_ have to be replicated in the other one, as their logic is exactly the same, and the same goes for the tests. 

only the type of cheque is the difference here—and not even at the structural level, just the name under which they are stored.

> He means that you're not holding any joint lock since the atomicity of their retrieval is supposed to be guaranteed by the state store rather than the `SWAP` code.

thanks for the clarification.

Reply:
> can you please point out what is wrong with the deep equal assertions? do you think we should have a custom method to compare these structs?

@mortelli I am at the opinion that we should reduce nesting levels in the codebase. This PR increases them.

My other consideration is that we're introducing a fragile contract that we'll need to maintain in the future. Explicit RPC calls mitigate this since you can mark a function for deprecation. How do you mark a key in a map for deprecation? i.e. tomorrow you want to announce that key number 1 is about to be deprecated/renamed in the next version/next year. How do you address this in the code? annotate a `const` as marked for deprecation? I personally find this not a clean approach to maintain exposed contracts on the long run. RPC functions with single responsibility make this easier and clearer to convey. It also reduces the nesting levels in the tests.

TBH I also find the current version easier to read and debug (for example, if someone from outside incentives track needs to debug a SWAP test).

Reply:
thank you @acud once again for your input, and thanks for approving.

i will make a note of all of this and make sure it's fairly discussed with @holisticode once he gets back. i promise to keep it in mind when closing #1445 and make any changes deemed necessary.